### PR TITLE
ARROW-11330: [Rust][DataFusion] add ExpressionVisitor to encode expression walking

### PR DESCRIPTION
**Pull request by alamb**
## Problem:
* There are several places in the DataFusion codebase where a walk of an Expression tree is needed
* The logic of how to walk the tree is replicated
* Adding new expression types often require many mechanically different but semantically the same changes in many places where no special treatment of such types is needed

This PR introduces a `ExpressionVisitor` trait and the `Expr::accept` function to consolidate this walking of the expression tree. It does not intend to change any functionality. 

If folks like this pattern, I have ideas for a similar type of trait `ExpressionRewriter` which can be used to rewrite expressions (much like `clone_with_replacement`) as a subsquent PR. I think this was mentioned by @Dandandan  in the [Rust roadmap](https://docs.google.com/document/d/1qspsOM_dknOxJKdGvKbC1aoVoO0M3i6x1CIo58mmN2Y/edit#heading=h.kstb571j5g5j)

cc @jorgecarleitao @Dandandan and @andygrove

### CODE REVIEW

File: rust/datafusion/src/optimizer/utils.rs
Comment (alamb) [inline rust/datafusion/src/optimizer/utils.rs:49]: This is a pretty good example of the kind of repetition that can be removed using this visitor pattern.
Note that I still left all expr types enumerated so that anyone who adds a new Expr type need to update this code, and (hopefully) think if they need to add special handling for that new expr types

**Comment by andygrove** [comment]
I took a quick look and this looks good to me. I will try and review in more detail tonight.

File: rust/datafusion/src/logical_plan/expr.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/logical_plan/expr.rs:426]:
```suggestion
    /// Performs a depth first walk of an expression and
```

File: rust/datafusion/src/logical_plan/expr.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/logical_plan/expr.rs:427]:
```suggestion
    /// its children, calling [`ExpressionVisitor::pre_visit`] and
```

**Review by jorgecarleitao — APPROVED** [review]
This is beautiful. Thank you, @alamb

**Comment by andygrove** [comment]
I just wanted to add that there is no need to wait for me to review before merging.

**Review by Dandandan — APPROVED** [review]
LGTM!
This should make it a bit less verbose / and error prone to write things like optimization rules.
I think we may just want to try to use it, if we find another pattern / approach works better we can adapt / improve it.

Performance-wise, it doesn't compile to a handwritten for loop of course (or even plain recursion), but for most foreseeable use cases, except some extreme examples, I think that should be fine for now.

**Comment by alamb** [comment]
Thanks @jorgecarleitao  and @Dandandan . I am personally quite excited at using this same idea for expression rewriting.

File: rust/datafusion/src/logical_plan/expr.rs
Comment (alamb) [inline rust/datafusion/src/logical_plan/expr.rs:543]:
```suggestion
/// `Expr::accept`, `ExpressionVisitor::visit` is invoked
```

**Pull request closed without merging**
